In [152]:
import pandas as pd
import numpy as np

In [153]:
pd.set_option('display.max_columns',None)

Importing the movie_metadata dataset

In [154]:
df = pd.read_csv('movie_metadata.csv')
df.head()

color      director_name  num_critic_for_reviews  duration  \
0  Color      James Cameron                   723.0     178.0   
1  Color     Gore Verbinski                   302.0     169.0   
2  Color         Sam Mendes                   602.0     148.0   
3  Color  Christopher Nolan                   813.0     164.0   
4    NaN        Doug Walker                     NaN       NaN   

   director_facebook_likes  actor_3_facebook_likes      actor_2_name  \
0                      0.0                   855.0  Joel David Moore   
1                    563.0                  1000.0     Orlando Bloom   
2                      0.0                   161.0      Rory Kinnear   
3                  22000.0                 23000.0    Christian Bale   
4                    131.0                     NaN        Rob Walker   

   actor_1_facebook_likes        gross                           genres  \
0                  1000.0  760505847.0  Action|Adventure|Fantasy|Sci-Fi   
1                 40000.0  309404152.0         Action|Adventure|Fantasy   
2                 11000.0  200074175.0        Action|Adventure|Thriller   
3                 27000.0  448130642.0                  Action|Thriller   
4                   131.0          NaN                      Documentary   

      actor_1_name                                        movie_title  \
0      CCH Pounder                                            Avatar    
1      Johnny Depp          Pirates of the Caribbean: At World's End    
2  Christoph Waltz                                           Spectre    
3        Tom Hardy                             The Dark Knight Rises    
4      Doug Walker  Star Wars: Episode VII - The Force Awakens    ...   

   num_voted_users  cast_total_facebook_likes          actor_3_name  \
0           886204                       4834             Wes Studi   
1           471220                      48350        Jack Davenport   
2           275868                      11700      Stephanie Sigman   
3          1144337                     106759  Joseph Gordon-Levitt   
4                8                        143                   NaN   

   facenumber_in_poster                                      plot_keywords  \
0                   0.0             avatar|future|marine|native|paraplegic   
1                   0.0  goddess|marriage ceremony|marriage proposal|pi...   
2                   1.0                bomb|espionage|sequel|spy|terrorist   
3                   0.0  deception|imprisonment|lawlessness|police offi...   
4                   0.0                                                NaN   

                                     movie_imdb_link  num_user_for_reviews  \
0  http://www.imdb.com/title/tt0499549/?ref_=fn_t...                3054.0   
1  http://www.imdb.com/title/tt0449088/?ref_=fn_t...                1238.0   
2  http://www.imdb.com/title/tt2379713/?ref_=fn_t...                 994.0   
3  http://www.imdb.com/title/tt1345836/?ref_=fn_t...                2701.0   
4  http://www.imdb.com/title/tt5289954/?ref_=fn_t...                   NaN   

  language country content_rating        budget  title_year  \
0  English     USA          PG-13  237000000.0$      2009.0   
1  English     USA          PG-13  300000000.0$      2007.0   
2  English      UK          PG-13  245000000.0$      2015.0   
3  English     USA          PG-13  250000000.0$      2012.0   
4      NaN     NaN            NaN          0.0$         NaN   

   actor_2_facebook_likes  imdb_score  aspect_ratio  movie_facebook_likes  
0                   936.0         7.9          1.78                 33000  
1                  5000.0         7.1          2.35                     0  
2                   393.0         6.8          2.35                 85000  
3                 23000.0         8.5          2.35                164000  
4                    12.0         7.1           NaN                     0

Basic EDA (Exploratory Data Analysis)

In [155]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5043 entries, 0 to 5042
Data columns (total 28 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   color                      5024 non-null   object 
 1   director_name              4939 non-null   object 
 2   num_critic_for_reviews     4993 non-null   float64
 3   duration                   5028 non-null   float64
 4   director_facebook_likes    4939 non-null   float64
 5   actor_3_facebook_likes     5020 non-null   float64
 6   actor_2_name               5030 non-null   object 
 7   actor_1_facebook_likes     5036 non-null   float64
 8   gross                      4159 non-null   float64
 9   genres                     5043 non-null   object 
 10  actor_1_name               5036 non-null   object 
 11  movie_title                5043 non-null   object 
 12  num_voted_users            5043 non-null   int64  
 13  cast_total_facebook_likes  5043 non-null   int64

The column 'budget', listed as an `object` type must be an `int` or `float` like data type, since it is a number.

In [156]:
#example
df['budget'][3]

'250000000.0$'

In [157]:
df['budget'] = df['budget'].str.replace('$','').\
                astype(float)
df['budget']

0       237000000.0
1       300000000.0
2       245000000.0
3       250000000.0
4               0.0
           ...     
5038            0.0
5039            0.0
5040         1400.0
5041            0.0
5042         1100.0
Name: budget, Length: 5043, dtype: float64

In [158]:
df.columns

Index(['color', 'director_name', 'num_critic_for_reviews', 'duration',
       'director_facebook_likes', 'actor_3_facebook_likes', 'actor_2_name',
       'actor_1_facebook_likes', 'gross', 'genres', 'actor_1_name',
       'movie_title', 'num_voted_users', 'cast_total_facebook_likes',
       'actor_3_name', 'facenumber_in_poster', 'plot_keywords',
       'movie_imdb_link', 'num_user_for_reviews', 'language', 'country',
       'content_rating', 'budget', 'title_year', 'actor_2_facebook_likes',
       'imdb_score', 'aspect_ratio', 'movie_facebook_likes'],
      dtype='object')

Creating a Dataframe copy for data manipulation. This is good practice when working with data as it allows you to make changes to the Dataframe while still maintaing the first one intact.

In [159]:
df_imdb = df.copy()

### 1.2 The duration column contains data on the film length. How many missing values are there in this column?

In [160]:
df_imdb['duration'].isna().value_counts()

duration
False    5028
True       15
Name: count, dtype: int64

If .isna() is True, then the row is missing a value. In this Dataframe we have 15 missing values for the 'duration' column


### 1.3. Replace the missing values in the duration column with the median value for this column.

1. Calculate the median value for the 'duration' column
2. Input the result in the Series with .fillna().

In [161]:
median_dur = df_imdb['duration'].median()
median_dur

103.0

In [162]:
df_imdb['duration'] = df_imdb['duration'].fillna(median_dur)
# checking for missing values after manipulation
df_imdb['duration'].isna().value_counts()

duration
False    5043
Name: count, dtype: int64

### 1.4. What is the average film length? Give the answer as a floating-point figure rounded to two decimal places.

In [163]:
avg_lenght = df_imdb['duration'].mean().round(2)

print(f'The average film lenght is {avg_lenght} minutes')

The average film lenght is 107.19 minutes


### 1.5. Create a movie_duration_category column, which will contain three categories depending on the film length:
• Category "1. <90" if the film is less than 90 minutes long

• Category "2. 90–120" if the film is between 90 minutes and two hours long (inclusively)

• Category "3. >120" if the film is more than two hours long


To do so, we are going to use the Numpy library function `np.where`, which allow us to use multiple if statements in a pd.Series. Since it is a 3 step process, we need to use the np.where twice - one time to identify movies with less than 90 minutes, and once again to distinguish between movies with more than 120 minutes and the movies between 90 and 120 minutes.

We could also use list comprehension or create a function and .apply it to the column. Doing so we miss the power of vectorization that Numpy offers it's users

In [164]:
df_imdb['movie_duration_category'] = np.where(
        df_imdb['duration'] < 90, '1. <90',
        np.where(df_imdb['duration'] > 120, '3. >120','2. >90-120')
)

In [165]:
df_imdb

color      director_name  num_critic_for_reviews  duration  \
0     Color      James Cameron                   723.0     178.0   
1     Color     Gore Verbinski                   302.0     169.0   
2     Color         Sam Mendes                   602.0     148.0   
3     Color  Christopher Nolan                   813.0     164.0   
4       NaN        Doug Walker                     NaN     103.0   
...     ...                ...                     ...       ...   
5038  Color        Scott Smith                     1.0      87.0   
5039  Color                NaN                    43.0      43.0   
5040  Color   Benjamin Roberds                    13.0      76.0   
5041  Color        Daniel Hsia                    14.0     100.0   
5042  Color           Jon Gunn                    43.0      90.0   

      director_facebook_likes  actor_3_facebook_likes      actor_2_name  \
0                         0.0                   855.0  Joel David Moore   
1                       563.0                  1000.0     Orlando Bloom   
2                         0.0                   161.0      Rory Kinnear   
3                     22000.0                 23000.0    Christian Bale   
4                       131.0                     NaN        Rob Walker   
...                       ...                     ...               ...   
5038                      2.0                   318.0     Daphne Zuniga   
5039                      NaN                   319.0     Valorie Curry   
5040                      0.0                     0.0     Maxwell Moody   
5041                      0.0                   489.0     Daniel Henney   
5042                     16.0                    16.0  Brian Herzlinger   

      actor_1_facebook_likes        gross                           genres  \
0                     1000.0  760505847.0  Action|Adventure|Fantasy|Sci-Fi   
1                    40000.0  309404152.0         Action|Adventure|Fantasy   
2                    11000.0  200074175.0        Action|Adventure|Thriller   
3                    27000.0  448130642.0                  Action|Thriller   
4                      131.0          NaN                      Documentary   
...                      ...          ...                              ...   
5038                   637.0          NaN                     Comedy|Drama   
5039                   841.0          NaN     Crime|Drama|Mystery|Thriller   
5040                     0.0          NaN            Drama|Horror|Thriller   
5041                   946.0      10443.0             Comedy|Drama|Romance   
5042                    86.0      85222.0                      Documentary   

         actor_1_name                                        movie_title  \
0         CCH Pounder                                            Avatar    
1         Johnny Depp          Pirates of the Caribbean: At World's End    
2     Christoph Waltz                                           Spectre    
3           Tom Hardy                             The Dark Knight Rises    
4         Doug Walker  Star Wars: Episode VII - The Force Awakens    ...   
...               ...                                                ...   
5038      Eric Mabius                           Signed Sealed Delivered    
5039      Natalie Zea                         The Following                
5040      Eva Boehnke                              A Plague So Pleasant    
5041        Alan Ruck                                  Shanghai Calling    
5042      John August                                 My Date with Drew    

      num_voted_users  cast_total_facebook_likes          actor_3_name  \
0              886204                       4834             Wes Studi   
1              471220                      48350        Jack Davenport   
2              275868                      11700      Stephanie Sigman   
3             1144337                     106759  Joseph Gordon-Levitt   
4                   8                        143                   N

### 1.6. Build a summary table for films released after 2000 (inclusively), to list the numbers of films:

• Table rows: year

• Table columns: movie duration category ("<90", "90–120", ">120")

• The year of release should be displayed in the YYYY format.

In [166]:
df_imdb['title_year'].isna().value_counts()

title_year
False    4935
True      108
Name: count, dtype: int64

In [167]:
# querying the Dataframe to only films released during and after the year 2000. Then, checking if still there are values missing

imdb_pivot = df_imdb.query('title_year>=2000').copy() #.copy() prevent pointer/memory errors being thrown after running.

imdb_pivot['title_year'] = imdb_pivot['title_year'].astype(int)
imdb_pivot['title_year'].isna().value_counts()

title_year
False    3597
Name: count, dtype: int64

In [168]:
imdb_pivot.head()

color      director_name  num_critic_for_reviews  duration  \
0  Color      James Cameron                   723.0     178.0   
1  Color     Gore Verbinski                   302.0     169.0   
2  Color         Sam Mendes                   602.0     148.0   
3  Color  Christopher Nolan                   813.0     164.0   
5  Color     Andrew Stanton                   462.0     132.0   

   director_facebook_likes  actor_3_facebook_likes      actor_2_name  \
0                      0.0                   855.0  Joel David Moore   
1                    563.0                  1000.0     Orlando Bloom   
2                      0.0                   161.0      Rory Kinnear   
3                  22000.0                 23000.0    Christian Bale   
5                    475.0                   530.0   Samantha Morton   

   actor_1_facebook_likes        gross                           genres  \
0                  1000.0  760505847.0  Action|Adventure|Fantasy|Sci-Fi   
1                 40000.0  309404152.0         Action|Adventure|Fantasy   
2                 11000.0  200074175.0        Action|Adventure|Thriller   
3                 27000.0  448130642.0                  Action|Thriller   
5                   640.0   73058679.0          Action|Adventure|Sci-Fi   

      actor_1_name                                movie_title  \
0      CCH Pounder                                    Avatar    
1      Johnny Depp  Pirates of the Caribbean: At World's End    
2  Christoph Waltz                                   Spectre    
3        Tom Hardy                     The Dark Knight Rises    
5     Daryl Sabara                               John Carter    

   num_voted_users  cast_total_facebook_likes          actor_3_name  \
0           886204                       4834             Wes Studi   
1           471220                      48350        Jack Davenport   
2           275868                      11700      Stephanie Sigman   
3          1144337                     106759  Joseph Gordon-Levitt   
5           212204                       1873          Polly Walker   

   facenumber_in_poster                                      plot_keywords  \
0                   0.0             avatar|future|marine|native|paraplegic   
1                   0.0  goddess|marriage ceremony|marriage proposal|pi...   
2                   1.0                bomb|espionage|sequel|spy|terrorist   
3                   0.0  deception|imprisonment|lawlessness|police offi...   
5                   1.0  alien|american civil war|male nipple|mars|prin...   

                                     movie_imdb_link  num_user_for_reviews  \
0  http://www.imdb.com/title/tt0499549/?ref_=fn_t...                3054.0   
1  http://www.imdb.com/title/tt0449088/?ref_=fn_t...                1238.0   
2  http://www.imdb.com/title/tt2379713/?ref_=fn_t...                 994.0   
3  http://www.imdb.com/title/tt1345836/?ref_=fn_t...                2701.0   
5  http://www.imdb.com/title/tt0401729/?ref_=fn_t...                 738.0   

  language country content_rating       budget  title_year  \
0  English     USA          PG-13  237000000.0        2009   
1  English     USA          PG-13  300000000.0        2007   
2  English      UK          PG-13  245000000.0        2015   
3  English     USA          PG-13  250000000.0        2012   
5  English     USA          PG-13  263700000.0        2012   

   actor_2_facebook_likes  imdb_score  aspect_ratio  movie_facebook_likes  \
0                   936.0         7.9          1.78                 33000   
1                  5000.0         7.1          2.35                     0   
2                   393.0         6.8          2.35                 85000   
3                 23000.0         8.5          2.35                164000   
5                   632.0         6.6          2.35                 24000   

  movie_duration_category  
0                 3. >120  
1                 3. >120  
2                 3. >120  
3                 3. >12

In [169]:
imdb_pivot = pd.pivot_table(imdb_pivot,values='movie_title',
                            columns='movie_duration_category',index='title_year',
                            aggfunc='count')

In [170]:
imdb_pivot

movie_duration_category  1. <90  2. >90-120  3. >120
title_year                                          
2000                         25         112       34
2001                         29         120       39
2002                         36         146       27
2003                         31         108       30
2004                         30         142       42
2005                         31         142       48
2006                         40         146       53
2007                         31         130       43
2008                         29         160       36
2009                         42         178       40
2010                         26         171       33
2011                         36         150       39
2012                         41         139       41
2013                         38         145       54
2014                         36         171       45
2015                         34         145       47
2016                          8          79       19

### 1.7. How many films between 90 minutes and two hours long were released in 2008?
A: 160 movies

In [171]:
imdb_pivot['2. >90-120']

title_year
2000    112
2001    120
2002    146
2003    108
2004    142
2005    142
2006    146
2007    130
2008    160
2009    178
2010    171
2011    150
2012    139
2013    145
2014    171
2015    145
2016     79
Name: 2. >90-120, dtype: int64

### 1.8. The plot_keywords column holds keywords characterizing the film's plot. Using the data in this column, create a column called movie_plot_category, to contain four categories depending on the key words in the column:

• Category "love_and_death" if the keywords include both "love" and "death"

• Category "love" if the keywords include the word "love"

• Category "death" if the keywords include the word "death"

• Category "other" if the keywords do not meet the conditions above

To do this categorization, we will use the `pd.str.contains()` function to map keywords and sort the movies accordingly. To do so, we will return to use the **df_imdb** Dataframe. 

Once again, we will use `np.where`, but a simpler solution with function will also be displayed.

In [172]:
df_imdb['plot_keywords']

0                  avatar|future|marine|native|paraplegic
1       goddess|marriage ceremony|marriage proposal|pi...
2                     bomb|espionage|sequel|spy|terrorist
3       deception|imprisonment|lawlessness|police offi...
4                                                     NaN
                              ...                        
5038               fraud|postal worker|prison|theft|trial
5039         cult|fbi|hideout|prison escape|serial killer
5040                                                  NaN
5041                                                  NaN
5042    actress name in title|crush|date|four word tit...
Name: plot_keywords, Length: 5043, dtype: object

In [173]:
df_imdb['movie_plot_category'] = np.where(
    #love_and_death
    df['plot_keywords'].str.contains("love") & df['plot_keywords'].str.contains("death"), "love_and_death",
    #love
    np.where(df['plot_keywords'].str.contains("love"), "love",
    #death
             np.where(df['plot_keywords'].str.contains("death"), "death", 
                      "other"))
             
             )

In [174]:
## using a function and .apply():

def categorize_keywords(keywords):
    if keywords.str.contains("love") and keywords.str.contains("death"):
        return "love_and_death"
    elif keywords.str.contains("love"):
        return "love"
    elif keywords.str.contains("death"):
        return "death"
    else:
        return "other"
#Not running, but then you use .apply() to 'literaly' apply the function to the column.
#df_imdb['movie_plot_category_2'] = df_imdb['plot_keywords'].apply(categorize_keywords)


In [175]:
df_imdb

color      director_name  num_critic_for_reviews  duration  \
0     Color      James Cameron                   723.0     178.0   
1     Color     Gore Verbinski                   302.0     169.0   
2     Color         Sam Mendes                   602.0     148.0   
3     Color  Christopher Nolan                   813.0     164.0   
4       NaN        Doug Walker                     NaN     103.0   
...     ...                ...                     ...       ...   
5038  Color        Scott Smith                     1.0      87.0   
5039  Color                NaN                    43.0      43.0   
5040  Color   Benjamin Roberds                    13.0      76.0   
5041  Color        Daniel Hsia                    14.0     100.0   
5042  Color           Jon Gunn                    43.0      90.0   

      director_facebook_likes  actor_3_facebook_likes      actor_2_name  \
0                         0.0                   855.0  Joel David Moore   
1                       563.0                  1000.0     Orlando Bloom   
2                         0.0                   161.0      Rory Kinnear   
3                     22000.0                 23000.0    Christian Bale   
4                       131.0                     NaN        Rob Walker   
...                       ...                     ...               ...   
5038                      2.0                   318.0     Daphne Zuniga   
5039                      NaN                   319.0     Valorie Curry   
5040                      0.0                     0.0     Maxwell Moody   
5041                      0.0                   489.0     Daniel Henney   
5042                     16.0                    16.0  Brian Herzlinger   

      actor_1_facebook_likes        gross                           genres  \
0                     1000.0  760505847.0  Action|Adventure|Fantasy|Sci-Fi   
1                    40000.0  309404152.0         Action|Adventure|Fantasy   
2                    11000.0  200074175.0        Action|Adventure|Thriller   
3                    27000.0  448130642.0                  Action|Thriller   
4                      131.0          NaN                      Documentary   
...                      ...          ...                              ...   
5038                   637.0          NaN                     Comedy|Drama   
5039                   841.0          NaN     Crime|Drama|Mystery|Thriller   
5040                     0.0          NaN            Drama|Horror|Thriller   
5041                   946.0      10443.0             Comedy|Drama|Romance   
5042                    86.0      85222.0                      Documentary   

         actor_1_name                                        movie_title  \
0         CCH Pounder                                            Avatar    
1         Johnny Depp          Pirates of the Caribbean: At World's End    
2     Christoph Waltz                                           Spectre    
3           Tom Hardy                             The Dark Knight Rises    
4         Doug Walker  Star Wars: Episode VII - The Force Awakens    ...   
...               ...                                                ...   
5038      Eric Mabius                           Signed Sealed Delivered    
5039      Natalie Zea                         The Following                
5040      Eva Boehnke                              A Plague So Pleasant    
5041        Alan Ruck                                  Shanghai Calling    
5042      John August                                 My Date with Drew    

      num_voted_users  cast_total_facebook_likes          actor_3_name  \
0              886204                       4834             Wes Studi   
1              471220                      48350        Jack Davenport   
2              275868                      11700      Stephanie Sigman   
3             1144337                     106759  Joseph Gordon-Levitt   
4                   8                        143                   N

### 1.9. The imdb_score column shows a viewer rating for the film. Build a table to reflect the average rating of films depending on which movie_plot_category category they belong to.

Using another method to create an grouping table between two columns is to use .groupby() and the desired calculation afterwards. We're also rounding up to two decimals.

In [176]:
df_avg_rating = pd.DataFrame(df_imdb.groupby('movie_plot_category')['imdb_score'].mean().round(2))
df_avg_rating

imdb_score
movie_plot_category            
death                      6.54
love                       6.39
love_and_death             6.51
other                      6.44

### 1.10. What is the average rating of films in the "love" category? Give the answer as a floating point figure rounded to two decimal places.

As shown above, the average rating in the 'love' category is 6.39.

### 1.11. The budget column contains the film's budget. What is the median budget for all the films listed? Give the answer as an integer.

Since we already fixed the 'budget' column values in the EDA process, we just have to convert it to an integer value.

In [177]:
df_imdb['budget']

0       237000000.0
1       300000000.0
2       245000000.0
3       250000000.0
4               0.0
           ...     
5038            0.0
5039            0.0
5040         1400.0
5041            0.0
5042         1100.0
Name: budget, Length: 5043, dtype: float64

In [178]:
median_budget = df_imdb['budget'].median()
print(f'The median budget for all listed films is {median_budget.astype(int)}')

The median budget for all listed films is 15000000
